### PR TITLE
Update rest-json guide to add HTTP filters

### PR DESCRIPTION
**Pull request by sarxos**
Hi @gsmet,

This is documentation update we discussed on Zulip in regards to the HTTP filters interceptors. The code update for quickstart is addressed by a separate PR available here https://github.com/quarkusio/quarkus-quickstarts/pull/212.

### CODE REVIEW

File: docs/src/main/asciidoc/rest-json-guide.adoc [hunk 1]
```diff
@@ -377,6 +377,64 @@ Let's do that and follow the same steps as before:
 
 This time, you can see our list of legumes.
 
+== HTTP filters and interceptors
+
+Both HTTP request and response can be intercepted by providing `ContainerRequestFilter` or `ContainerResponseFilter`
+implementations respectively. These filters are suitable for processing the metadata associated with a message: HTTP
+headers, query parameters, media type, and other metadata. They also have the capability to abort a REST invocation
```
Comment (cescoffier) [inline docs/src/main/asciidoc/rest-json-guide.adoc]: The word REST is not great here. Maybe something like: 

```
They also have the capability to abort the request processing, for instance when the user does not have the permissions to access the endpoint.
```

WDYT?

File: docs/src/main/asciidoc/rest-json-guide.adoc [hunk 2]
```diff
@@ -321,7 +321,7 @@ import javax.ws.rs.core.Response;
 @Consumes(MediaType.APPLICATION_JSON)
 public class LegumeResource {
 
-    private Set<Legume> legumes = Collections.newSetFromMap(Collections.synchronizedMap(new LinkedHashMap<>()));
+    private Set<Legume> legumes = Collections.synchronizedSet(new LinkedHashSet<>());
```
Comment (cescoffier) [inline docs/src/main/asciidoc/rest-json-guide.adoc]: Any reason to change this?

**Review by cescoffier — DISMISSED** [review]

**Comment by cescoffier** [comment]
Mostly good, I've made a few comments.

File: docs/src/main/asciidoc/rest-json-guide.adoc (hunk 2 above)
Comment (sarxos) [reply docs/src/main/asciidoc/rest-json-guide.adoc]: No one is constructing synchronized set this way :) If you want concurrent access then the the correct way is:

```java
Collections.newSetFromMap(new ConcurrentHashMap<>());
```

And if you want synchronized access then it should be done like this:

```java
Collections.synchronizedSet(new LinkedHashSet<>());
```

Having:

```java
Collections.newSetFromMap(Collections.synchronizedMap(new LinkedHashMap<>()));
```

Is an exaggerated use of `Collections` API.

File: docs/src/main/asciidoc/rest-json-guide.adoc (hunk 1 above)
Comment (sarxos) [reply docs/src/main/asciidoc/rest-json-guide.adoc]: Yes, what you proposed is much better. Will change this sentence.

Comment (sarxos) [reply docs/src/main/asciidoc/rest-json-guide.adoc]: Done.

File: docs/src/main/asciidoc/rest-json-guide.adoc [hunk 3]
```diff
@@ -377,6 +377,64 @@ Let's do that and follow the same steps as before:
...
 
+== HTTP filters and interceptors
+
+Both HTTP request and response can be intercepted by providing `ContainerRequestFilter` or `ContainerResponseFilter`
+implementations respectively. These filters are suitable for processing the metadata associated with a message: HTTP
+headers, query parameters, media type, and other metadata. They also have the capability to abort the request
+processing, for instance when the user does not have the permissions to access the endpoint.
+
+Let's use `ContainerRequestFilter` to add logging capability to our service. We can do that by implementing
+`ContainerRequestFilter` and annotating it with `@Provider` annotation:
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-json-guide.adoc]: I would favor `with *the* @Provider annotation`.

File: docs/src/main/asciidoc/rest-json-guide.adoc [hunk 4]
```diff
@@ -377,6 +377,64 @@ Let's do that and follow the same steps as before:
...
+        final String method = context.getMethod();
+        final String path = info.getPath();
+        final String address = request.getRemoteAddr();
+
+        LOG.infof("Request %s %s from IP %s", method, path, address);
+    }
+}
+----
+
+Now, whenever REST method is invoked, the request will be logged into the console:
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-json-guide.adoc]: whenever **a** REST method

File: docs/src/main/asciidoc/rest-json-guide.adoc (hunk 2 above)
Comment (gsmet) [reply docs/src/main/asciidoc/rest-json-guide.adoc]: Agreed that I was not on my best day when I wrote this :).

**Review by gsmet — CHANGES_REQUESTED** [review]
@sarxos I added a few minor comments. Could you fix them and squash everything?

Very nice addition, thanks!

File: docs/src/main/asciidoc/rest-json-guide.adoc (hunk 3 above)
Comment (sarxos) [reply docs/src/main/asciidoc/rest-json-guide.adoc]: Done.

File: docs/src/main/asciidoc/rest-json-guide.adoc (hunk 4 above)
Comment (sarxos) [reply docs/src/main/asciidoc/rest-json-guide.adoc]: Done.

**Comment by sarxos** [comment]
@gsmet @cescoffier All comments addressed.

**Review by gsmet — APPROVED** [review]
Perfect, nice work!

**Pull request merged**